server/src/main/java/umm3601/todo/TodoDatabase.java

```diff
@@ -0,0 +1,159 @@
+        filteredTodos = filterTodosByStatus(filteredTodos, false);
+      }
+
+    }
+
+    // Process other query parameters here...
+    if (queryParams.containsKey("category")) {
+      String categoryParam = queryParams.get("category").get(0);
+      filteredTodos = filterTodosByCategory(filteredTodos, categoryParam);
+    }
+
+    // Filter by matching string in body
+    if (queryParams.containsKey("contains")) {
+      String containsParam = queryParams.get("contains").get(0);
+      filteredTodos = filterTodosByContains(filteredTodos, containsParam);
+    }
+
+    // Limit number of todos displayed.
+    if (queryParams.containsKey("limit")) {
+      String limitParam = queryParams.get("limit").get(0);
+      try {
+        int limitTarget = Integer.parseInt(limitParam);
+        filteredTodos = limitTodos(filteredTodos, limitTarget); // This is a placeholder, we need to
+        // find a way to limit the number of todos shown.
+      } catch (NumberFormatException e) {
+        throw new BadRequestResponse("Specified limit '" + limitParam + "' can't be parsed to an integer");
+      }
+    }
+
+    return filteredTodos;
+  }
+
+  /**
+   * Get an array of all the todos having the target owner.
+   *
+   * @param todos     the list of todos to filter by owner
+   * @param targetOwner the target owner to look for
+   * @return an array of all the todos from the given list that have the target
+   *         owner
+   */
+  public Todo[] filterTodosByOwner(Todo[] todos, String targetOwner) {
+    return Arrays.stream(todos).filter(x -> x.owner.equals(targetOwner)).toArray(Todo[]::new);
+  }
+
+  /**
+   * Get an array of all the todos having the target status.
+   *
+   * @param todos         the list of todos to filter by status
+   * @param targetStatus the target status to look for
+   * @return an array of all the todos from the given list that have the target
+   *         status
+   */
+  public Todo[] filterTodosByStatus(Todo[] todos, boolean targetStatus) {
+    return Arrays.stream(todos).filter(x -> x.status == targetStatus).toArray(Todo[]::new);
+  }
+
+  /**
+   * Get an array of all the todos having the target category.
+   *
+   * @param todos         the list of todos to filter by category
+   * @param targetCategory the target category to look for
+   * @return an array of all the todos from the given list that have the target
+   *         category
+   */
+  public Todo[] filterTodosByCategory(Todo[] todos, String targetCategory) {
+    return Arrays.stream(todos).filter(x -> x.category.equals(targetCategory)).toArray(Todo[]::new);
+  }
+
+  /**
+   * Get an array of all the todos containing the targetString.
+   *
+   * @param todos         the list of todos to filter by category
+   * @param targetString the target category to look for
+   * @return an array of all the todos from the given list that contain the target
+   *         string
+   */
+  public Todo[] filterTodosByContains(Todo[] todos, String targetString) {
```

Comment:
This needs to be a case sensitive method.  In your tests, you search for "Ipsum", which only has 12 instances in the `todos.json`, and if you were to search for "ipsum" you would get 71, and case-insensitive, 83 total instances of "Ipsum" or "ipsum".

Reply:
Ok sorry about that. I was under the impression that searching by a string wouldn't be case sensitive since the user would be searching for the content, whether they capitalize the first letter or no. I'll remove the .toLowerCase() methods in the the stream lambda